src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -710,6 +715,12 @@ ExecuteSingleModifyTask(CitusScanState *scanState, Task *task, bool expectResult
 										  taskRequiresTwoPhaseCommit,
 										  startedInTransaction);
 
+	/*
+	 * If we are dealing with partitioned table, we also need to lock its
+	 * partitions.
+	 */
+	LockPartitionRelationsIfPartitioned(relationId, RowExclusiveLock);
```

Comment:
I also think that we should take the locks. However, doing it in this PR would make things a lot complicated. We'd need to add `relationShardList` to `Job` and acquire the locks on the executor (or smth like that). 

I think we should open an issue and acquire the locks for `SELECT`s properly on the coordinator node. (Now that we acquire `AccessShareLock` just because deparsing acquires it.) @marcocitus What do you think?